keps/sig-node/4603-tune-crashloopbackoff/README.md

```diff
@@ -0,0 +1,1222 @@
+the heat death of the universe as depicted below:
+
+![A graph showing the backoff decay for a Kubernetes pod in
+CrashLoopBackoff](./crashloopbackoff-succeedingcontainer.png "CrashLoopBackoff
+decay")
+
+Remember that the backoff counter is reset if containers run longer than 10
+minutes, so in the worst case where a container always exits after 9:59:59, this
+means in the first 30 minute period, the container will restart twice. In a more
+fruitful example, for a fast exiting container crashing every 10 seconds or so,
+in the first 30 minutes the container will restart about 10 times, with the
+first four restarts in the first 5 minutes.
+
+This KEP proposes changing the existing backoff curve to load more restarts
+earlier by changing the initial value of the exponential backoff. In an effort
+to anticipate API server stability ahead of the experiential data we can collect
+during alpha, the proposed changes are to both reduce the initial value, and
+step function to a higher delay cap once the decay curve triggers the same
+number of total restarts as experienced today in a 10 minute time horizon, in
+order to approximate load (though not rate) of pod restart API server requests.
+
+The detailed methodology for determining the implementable starting value and
+step function cap, and benchmarking it during and after alpha, is enclosed in
+Design Details. In short, the current proposal is to implement a new initial
+value of 1s, and a catch-up delay of 569 seconds (almost 9.5 minutes) on the 6th
+retry.
+
+!["A graph showing the delay interval function needed to maintain restart
+number"](controlfornumberofrestarts.png
+"Alternate CrashLoopBackoff decay")
+
+
+#### Flat-rate restarts for `Success` (exit code 0)
+
+We start from the assumption that exit code 0, "Success", means that a workload
+completed as expected.
+
+The wording of the public documentation
+([ref](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#container-restarts))
+and the naming of the `CrashLoopBackOff` state itself implies that it is a
+remedy for a container not exiting as intended, but the restart delay decay
+curve is applied to even successful pods if their `restartPolicy = Always`. On the
+canonical tracking issue for this problem, a significant number of requests
+focus on how an exponential decay curve is inappropriate for workloads
+completing as expected, and unnecessarily delays healthy workloads from being
+rescheduled.
+
+This KEP intends to vastly simplify and cut down on the restart delays for
+workloads completing as expected, as detectable by their exit code 0. The target
+is to get as close to the capacity of kubelet to instantly restart as possible,
+anticipated to be somewhere within 0-10s flat rate + jitter delay for each
+restart. The detailed methodology for determining the implementable starting
+value and benchmarking it before and during alpha is in the Design Details
+section, and for the first pass will be set to 5s.
```

Comment:
Notes from IRL convo for visibility/posterity, will integrate into KEP before resolving:

Yes, the first few restarts of a failed container will be faster than a successful container. I still think this is correct because the goal of restarting failed containers is to take maximum advantage of quickly recoverable situations, while the goal of restarting successful containers is only to get them to run again sometime and not penalize them with longer waits later when they've behaving as expected.

Reply:
Added for posterity in alternatives section via [7de94ab](https://github.com/kubernetes/enhancements/pull/4604/commits/7de94ab686213b2f3559348dd0d02fecadd0992c)